### PR TITLE
Add new fido_init_with_log() to allow adding customized logger

### PR DESCRIPTION
**Pull request by ikimonogakari**
Hi Pedro,

I messed up with merging the master's change, so I opened a new pull request regarding the old PR#96. Please take another look.

This function should not be called together with fido_init(), which only register the default logger.

Thanks,
Cary

### CODE REVIEW

File: src/fido/types.h [hunk 1]
```diff
@@ -34,6 +35,10 @@ typedef enum {
 	FIDO_OPT_TRUE,     /* explicitly set option to true */
 } fido_opt_t;
 
+// log function pointer type
+// Caller would call va_start and va_end to the va_list argument.
+typedef void (*fido_log_fp_t)(const char *, va_list);
```
[the remark this answers is not in the document]
Comment (martelletto) [reply src/fido/types.h]: In that case, fido_log_handler_t would be a more descriptive name.

File: src/dev.c [hunk 2]
```diff
@@ -295,7 +295,14 @@ void
 fido_init(int flags)
 {
 	if (flags & FIDO_DEBUG || getenv("FIDO_DEBUG") != NULL)
-		fido_log_init();
+		fido_log_init(NULL);
+}
+
+void
+fido_init_with_logger(int flags, fido_log_fp_t log_fp)
```
[the remark this answers is not in the document]
Comment (ikimonogakari) [reply src/dev.c]: Agreed and changed.

File: src/fido/types.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikimonogakari) [reply src/fido/types.h]: I made a max length of a single log line 5000 bytes, which is a constant.
Is that fine, or do you have any better suggestions?

File: CMakeLists.txt [hunk 3]
```diff
@@ -114,6 +114,7 @@ else()
 		# MinGW is stuck with a flavour of C89.
 		add_definitions(-DFIDO_NO_DIAGNOSTIC)
 		add_definitions(-DWC_ERR_INVALID_CHARS=0x80)
+		set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-parameter")
```
[the remark this answers is not in the document]
Comment (ikimonogakari) [reply CMakeLists.txt]: Yes.
In the fido_set_log_handler function.
The log_fp parameter is only used in the if block, and there is no else block any more.
That is considered as unused parameters in mingw.

File: src/dev.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (martelletto) [reply src/dev.c]: Seems consistent to me.

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (martelletto) [reply CMakeLists.txt]: Thanks.

File: src/fido/types.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (martelletto) [reply src/fido/types.h]: A fixed limit is fine.

**Pull request merged**